ast/check.go

```diff
@@ -146,7 +146,81 @@ func (tc *typeChecker) checkLanguageBuiltins(env *TypeEnv, builtins map[string]*
 	return env
 }
 
+func getPrefixToOverride(name string, env *TypeEnv) (string, types.Type) {
+	t := env.tree.Get(MustParseRef(name))
+	if t != nil {
+		return name, t
+	}
+	names := strings.Split(name, ".")
+	if len(names) > 1 {
+		prefix, t := getPrefixToOverride(strings.TrimSuffix(name, "."+names[len(names)-1]), env)
+		if t != nil {
+			return prefix, t
+		}
+	}
+	return "", nil
+}
+
+func override(names []string, t types.Type, o types.Type) types.Type {
+	obj := t.(*types.Object)
+	staticProps := obj.StaticProperties()
+	newStaticProps := []*types.StaticProperty{}
+
+	for _, prop := range staticProps {
+		if len(names) == 1 && names[0] == prop.Key.(string) {
+			prop.Value = o
+		} else if len(names) > 0 && names[0] == prop.Key.(string) {
+			prop.Value = override(names[1:], prop.Value, o)
+		}
+		newStaticProps = append(newStaticProps, types.NewStaticProperty(prop.Key, prop.Value))
+	}
+	return types.NewObject(newStaticProps, nil)
+}
+
+func (tc *typeChecker) processAnnotation(env *TypeEnv, rule *Rule) {
+	// If rule has a schema annotation, then process the schema and add it to TypeEnv
+	// Annotations must immediately precede the rule definition
+	// They are of the form: #@rulesSchema=<expr>:<schema-key>
+
+	if rule.Annotation != nil {
+		errors := []*Error{}
+		for _, annot := range rule.Annotation {
+			if env.schemaSet == nil || env.schemaSet.ByPath == nil {
+				errors = append(errors, NewError(TypeErr, rule.Location, "Schemas need to be supplied for the annotation: %s", annot.Schema))
+				break
+			}
+			schema := env.schemaSet.ByPath[annot.Schema]
+			if schema == nil {
```

Comment:
[nit] I think the following is a little more common:
```suggestion
			schema, ok := env.schemaSet.ByPath[annot.Schema]
			if !ok {
```

Reply:
Done (commit https://github.com/open-policy-agent/opa/pull/3123/commits/54395f8c2cd0e71df407142a2ee8f223092534f9)